### DISCUSSION TITLE
Can I use a custom shape function using numpy or torch?

### DISCUSSION THREAD

**Discussion by narnia-ai-mason** [libfive/libfive#555] [Q&A] [answered]
Hi, thank you for developing such an awesome software.
Especially, I really like the Python binding feature.
btw, can I use a custom python shape function using other libraries, such as numpy and torch?
My custom function can provide Hermit data (i.e., the signed distance value and its gradient for normal vectors).
For example, 

```Python
class MyShape(nn.Module):
    def __init__(self, *args, **kwargs):
        super().__init__()

        # model definitions here
        ...

    def forward(self, x, y, z):
        xyz = torch.cat([x, y, z], dim=1)

        signed_distance = self.model(xyz)
        normals = self.compute_normal(xyz)

        return signd_distance, normals

    def compute_normal(self, xyz):
        # normal vectors can easily be computed by using autodiff feature in torch
        ...
        return normals
```

If it is not possible for now, are you willing to possibly add this feature? What do you think of this?

Thanks :)

**Comment by mkeeter** [comment, marked as answer]
Unfortunately, that's not possible right now.

`libfive` works by compiling shapes down to math trees, usually using evaluation tracing (that's how the [`@shape`](https://github.com/libfive/libfive/blob/master/libfive/bind/python/libfive/shape.py#L278) decorator works).  These tree objects are what's rendered, so it doesn't support arbitrary code.

If you could export such a math tree from Torch, that would be one option; another option would be to use the Oracle API (which requires writing C++, discussed [here](https://github.com/libfive/libfive/discussions/468)).  I don't have the resources to do either implementation myself.

**Reply by narnia-ai-mason** [reply]
Oh I got it. Then I have to figure out how to convert torch's computational graph to libfive math tree. Right?
